crates/ty_python_semantic/src/types/function.rs

```diff
@@ -882,6 +883,95 @@ impl<'db> FunctionType<'db> {
     }
 }
 
+/// Evaluate an `isinstance` call. Return `Truthiness::AlwaysTrue` if we can definitely infer that
+/// this will return `True` at runtime, `Truthiness::AlwaysFalse` if we can definitely infer
+/// that this will return `False` at runtime, or `Truthiness::Ambiguous` if we should infer `bool`
+/// instead.
+fn is_instance_truthiness<'db>(
+    db: &'db dyn Db,
+    ty: Type<'db>,
+    class: ClassLiteral<'db>,
+) -> Truthiness {
+    let is_instance = |ty: &Type<'_>| {
+        if let Type::NominalInstance(instance) = ty {
+            if instance
+                .class
+                .is_subclass_of(db, ClassType::NonGeneric(class))
+            {
+                return true;
+            }
+        }
+        false
+    };
+
+    let always_true_if = |test: bool| {
+        if test {
+            Truthiness::AlwaysTrue
+        } else {
+            Truthiness::Ambiguous
+        }
+    };
+
+    match ty {
+        Type::Union(_) => {
+            // We do not handle unions specifically here, because something like `A | SubclassOfA` would
+            // have been simplified to `A` anyway
+            Truthiness::Ambiguous
+        }
+        Type::Intersection(intersection) => always_true_if(
+            intersection
+                .positive(db)
+                .iter()
+                .any(|element| is_instance_truthiness(db, *element, class).is_always_true()),
+        ),
+
+        Type::NominalInstance(_) => always_true_if(is_instance(&ty)),
+
+        Type::BooleanLiteral(_)
+        | Type::BytesLiteral(_)
+        | Type::IntLiteral(_)
+        | Type::StringLiteral(_)
+        | Type::LiteralString
+        | Type::ModuleLiteral(_)
+        | Type::EnumLiteral(_) => always_true_if(
+            ty.literal_fallback_instance(db)
+                .as_ref()
+                .is_some_and(is_instance),
+        ),
+
+        Type::Tuple(..) => always_true_if(class.is_known(db, KnownClass::Tuple)),
+
+        Type::FunctionLiteral(..) => {
+            always_true_if(is_instance(&KnownClass::FunctionType.to_instance(db)))
+        }
```

Comment:
I think we can optimize the same way for function-literals as you did for tuples?

```suggestion
        Type::Tuple(..) => always_true_if(class.is_known(db, KnownClass::Tuple)),

        Type::FunctionLiteral(..) => {
            always_true_if(class.is_known(db, KnownClass::FunctionType))
        }
```

Reply:
I think the existing check for function literals is more general. It would also work for `isinstance(some_function, object)`. I had to rewrite the one for tuples because it didn't work the other way around (due to the representation of tuple instances, I think)

Reply:
Aha, yet another reason for me to remove `Type::Tuple` entirely!